### PR TITLE
Create new Uri struct

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Here is the first draft of the new `Uri` struct. If this seems okay to you @seanmonstar, I'll remove the actual `UriRequest` and replace it with the new `Uri`.

PS: This is a WIP so do not merge!

### CODE REVIEW

**Comment by seanmonstar** [comment]
This is absolutely the direction I had in mind. I would also remove the use
of the Url type here, since it actually does more than is wanted on the
server side, and does several allocations. Since this type is created for
every request, we need to make it faster.

I can do that as follow up if you prefer.

**Comment by GuillaumeGomez** [comment]
I think I can do it in a second step. I'll just "borrow" the url method and use them directly instead of creating a type. For now I'll replace `RequestUri` everywhere and then I'll remove the url type usage.

File: examples/server.rs [hunk 1]
```diff
@@ -23,12 +23,12 @@ impl Service for Echo {
 
     fn call(&mut self, req: Request) -> Self::Future {
         ::futures::finished(match (req.method(), req.path()) {
-            (&Get, Some("/")) | (&Get, Some("/echo")) => {
+            (&Get, "/") | (&Get, "/echo") => {
```
Comment (seanmonstar) [inline examples/server.rs]: This looks so much better!

File: src/uri.rs [hunk 2]
```diff
@@ -1,11 +1,260 @@
...
+
+    pub fn fragment(&self) -> Option<&str> {
+        if let Some(len) = self.fragment {
+            Some(&self.source[self.source.len() - len..])
+        } else {
+            None
+        }
+    }
+
+    pub fn uri(&self) -> &str {
```
Comment (seanmonstar) [inline src/uri.rs]: I'd probably replace this with an `AsRef<str>` impl, since it sounds like stuttering when saying `req.uri().uri()`.

Comment (GuillaumeGomez) [inline src/uri.rs]: Hum, good point!

**Comment by GuillaumeGomez** [comment]
I think I'll stop here for this PR. I cleaned up a bit and remove the old RequestUri.

@seanmonstar: Do you see anything I might have forgotten?

**Pull request closed without merging**